docs/spec.md

```diff
@@ -173,10 +190,56 @@ values of type `tensor<T>`).
   and an **imaginary part** of the same **element type**. Supported complex
   types are `complex<f32>` (both parts are of type `f32`) and `complex<f64>`
   (both parts are of type `f64`).
-* In the future, we are also planning to introduce **quantized types** that
-  represent integer values obtained via uniform quantization of floating-point
-  values using given scales and zero points
-  ([#588](https://github.com/openxla/stablehlo/issues/588)).
+* **Quantized types** represent integer values of a **storage type** in the
+  range from `storage_min` to `storage_max` (inclusive) that correspond to
+  floating-point values of an **expressed type**. For a given integer value `i`,
+  the corresponding floating-point value `f` can be computed as
+  `f = (i - zero_point) * scale`, where `scale` and `zero_point` are called
+  **quantization parameters**. The `storage_min` and `storage_max`, being
+  optional, have default values which should satisfy the associated constraints
+  as laid out in this document.
+
+Quantization can be **per-tensor**, meaning, having one `scale` and/or
+`zero_point` for the entire tensor or can be **per-axis**, meaning, having
+multiple `scales` and/or `zero_points`, one pair per slice of a particular
+dimension `quantized_dimension`. More formally, in a tensor `t` of per-axis
+quantized element type, there are `dim(t, quantized_dimension)` slices of the
+`quantized_dimension`: `t[:, ..., 0, ..., :], t[:, ..., 1, ..., :]`, etc. All
+elements in the `i`th slice use `scales[i]` and `zero_points[i]` as their
+quantization parameters.
+
+Quantized types satisfy the following constraints:
+
+* (C1) `num_bits(storage_type) < num_bits(expressed_type)`.
+* (C2) `type(storage_min) = storage_type`.
+* (C3) `type(storage_max) = storage_type`.
+* (C4) If `storage_min` is empty and `d = num_bits(storage_type)`:
+  * `storage_min = is_signed(storage_type) ? -2^(d-1) : 0`.
+* (C5) If `storage_max` is empty:
+  * `storage_max = is_signed(storage_type) ? 2^(d-1) : 2^d - 1`.
+* (C6) If `storage_min` is not empty:
+  * If `is_signed(storage_type)`, `storage_min >= -2^(d-1)`.
+  * If `is_unsigned(storage_type)`, `storage_min >= 0`.
```

Comment:
Can we abstract this away by using `min_value(type)` and `max_value(type)` metafunctions?